clientd3d/d3drender.c

```diff
@@ -7492,27 +7550,51 @@ void D3DRenderOverlaysDraw(d3d_render_pool_new *pPool, room_type *room, Draw3DPa
 							if (ROOM_OVERRIDE_DEPTH1 & GetRoomFlags())
 							{
 								depthf = GetOverrideRoomDepth(SF_DEPTH1);
+                        bUsingAlternateDepth = TRUE;
 							}
 						break;
 						case SF_DEPTH2:
 							if (ROOM_OVERRIDE_DEPTH2 & GetRoomFlags())
 							{
 								depthf = GetOverrideRoomDepth(SF_DEPTH2);
+                        bUsingAlternateDepth = TRUE;
 							}
 						break;
 						case SF_DEPTH3:
 							if (ROOM_OVERRIDE_DEPTH3 & GetRoomFlags())
 							{
 								depthf = GetOverrideRoomDepth(SF_DEPTH3);
+                        bUsingAlternateDepth = TRUE;
 							}
 						break;
 						}
 					}
 
 					MatrixRotateY(&rot, (float)angleHeading * 360.0f / 4096.0f * PI / 180.0f);
 					MatrixTranspose(&rot, &rot);
-					MatrixTranslate(&mat, (float)pRNode->motion.x, (float)max(bottom,
-						pRNode->motion.z) - depthf, (float)pRNode->motion.y);
+
+               if (pRNode->obj.flags & OF_HANGING)
+               {
+   					MatrixTranslate(&mat, (float)pRNode->motion.x, top - pRNode->obj.boundingHeight, (float)pRNode->motion.y);
+               }
+               else if (bUsingAlternateDepth)
+					{
+                  if (pRNode->obj.flags & OF_BOUNCING)
+                  {
+                     MatrixTranslate(&mat, (float)pRNode->motion.x, (float)max(depth,
+                        pRNode->motion.z), (float)pRNode->motion.y);
+                  }
+                  else
+                  {
+   					   MatrixTranslate(&mat, (float)pRNode->motion.x, depthf, (float)pRNode->motion.y);
+                  }
+					}
+               else
+               {
+   					MatrixTranslate(&mat, (float)pRNode->motion.x, (float)max(bottom,

```

Comment:
As before, use a yoffset variable so that MatrixTranslate can be called only once.